File: scrape_moves.js
const fs = require('fs');

async function scrapeMoves() {
    console.log("Starting PokeAPI Move Scraper...");
    try {
        // 1. Get the list of all moves
        // Note: As of Gen 9, there are around 900+ moves. We fetch up to 1000 to be safe.
        const listResponse = await fetch('https://pokeapi.co/api/v2/move?limit=1000');
        const listData = await listResponse.json();
        const moveUrls = listData.results.map(m => m.url);

        console.log(`Found ${moveUrls.length} moves to process.`);

        const movesDatabase = [];
        let count = 0;

        // Process in batches so we don't overwhelm the API
        const batchSize = 50;
        for (let i = 0; i < moveUrls.length; i += batchSize) {
            const batchUrls = moveUrls.slice(i, i + batchSize);
            const batchPromises = batchUrls.map(async (url) => {
                try {
                    const res = await fetch(url);
                    if (!res.ok) return null;
                    const moveData = await res.json();

                    // Extract Korean name
                    const krNameObj = moveData.names.find(n => n.language.name === 'ko');
                    if (!krNameObj) return null; // Skip if no Korean translation exists

                    const name = krNameObj.name;

                    // Extract Korean flavor text (preferably from the latest version)
                    let flavorText = "";
                    const krFlavorTexts = moveData.flavor_text_entries.filter(f => f.language.name === 'ko');
                    if (krFlavorTexts.length > 0) {
                        // Get the last one, which is usually the newest generation
                        flavorText = krFlavorTexts[krFlavorTexts.length - 1].flavor_text;
                        // Clean up control characters
                        flavorText = flavorText.replace(/[\n\f\r]/g, ' ');
                    }

                    // Extract Type
                    let type = "노말"; // Default
                    if (moveData.type) {
                        const typeUrl = moveData.type.url;
                        const typeRes = await fetch(typeUrl);
                        if (typeRes.ok) {
                            const typeData = await typeRes.json();
                            const krTypeObj = typeData.names.find(n => n.language.name === 'ko');
                            if (krTypeObj) type = krTypeObj.name;
                        }
                    }

                    // Extract Damage Class (Category)
                    let category = "물리"; // Default
                    if (moveData.damage_class) {
                        const className = moveData.damage_class.name;
                        if (className === 'physical') category = '물리';
                        else if (className === 'special') category = '특수';
                        else if (className === 'status') category = '변화';
                    }

                    return {
                        name: name,
                        type: type,
                        category: category,
                        power: moveData.power || "-",
                        accuracy: moveData.accuracy || "-",
                        pp: moveData.pp || "-",
                        description: flavorText
                    };
                } catch (err) {
                    console.error(`Failed to process URL ${url}`, err.message);
                    return null;
                }
            });

            const batchResults = await Promise.all(batchPromises);
            for (const result of batchResults) {
                if (result) {
                    movesDatabase.push(result);
                }
            }

            count += batchResults.length;
            console.log(`Processed ${count}/${moveUrls.length} moves...`);
        }

        console.log(`Successfully scraped ${movesDatabase.length} moves.`);

        fs.writeFileSync('moves.json', JSON.stringify(movesDatabase, null, 2), 'utf8');
        console.log("Saved to moves.json!");

    } catch (error) {
        console.error("Fatal Error during scraping:", error);
    }
}

scrapeMoves();

File: scrape_moves.py
import urllib.request
import json
import ssl
import sys

def fetch_json(url):
    ctx = ssl.create_default_context()
    ctx.check_hostname = False
    ctx.verify_mode = ssl.CERT_NONE
    req = urllib.request.Request(url, headers={'User-Agent': 'Mozilla/5.0'})
    with urllib.request.urlopen(req, context=ctx) as response:
        return json.loads(response.read().decode('utf-8'))

def scrape():
    print("Fetching type name mappings...")
    type_map = {}
    type_list = fetch_json('https://pokeapi.co/api/v2/type?limit=30')
    for t in type_list['results']:
        td = fetch_json(t['url'])
        kr_names = [n['name'] for n in td['names'] if n['language']['name'] == 'ko']
        if kr_names:
            type_map[t['name']] = kr_names[0]
            
    print("Fetching move list...")
    list_data = fetch_json('https://pokeapi.co/api/v2/move?limit=1000')
    urls = [m['url'] for m in list_data['results']]
    print(f"Found {len(urls)} moves.")
    
    moves = []
    count = 0
    for url in urls:
        try:
            data = fetch_json(url)
            
            kr_names = [n['name'] for n in data['names'] if n['language']['name'] == 'ko']
            if not kr_names:
                continue
            name = kr_names[0]
            
            kr_flavor = [f['flavor_text'] for f in data['flavor_text_entries'] if f['language']['name'] == 'ko']
            flavor_text = kr_flavor[-1].replace('\n', ' ').replace('\f', ' ').replace('\r', ' ') if kr_flavor else ""
            
            type_name = "노말"
            if data['type'] and data['type']['name'] in type_map:
                type_name = type_map[data['type']['name']]
            
            cat = "물리"
            if data['damage_class']:
                c_name = data['damage_class']['name']
                if c_name == 'physical': cat = '물리'
                elif c_name == 'special': cat = '특수'
                elif c_name == 'status': cat = '변화'
                
            moves.append({
                "name": name,
                "type": type_name,
                "category": cat,
                "power": data['power'] if data['power'] is not None else "-",
                "accuracy": data['accuracy'] if data['accuracy'] is not None else "-",
                "pp": data['pp'] if data['pp'] is not None else "-",
                "description": flavor_text
            })
            count += 1
            if count % 50 == 0:
                print(f"Processed {count} moves...")
        except Exception as e:
            print(f"Error skipping {url}: {e}")
            
    with open('moves.json', 'w', encoding='utf-8') as f:
        json.dump(moves, f, ensure_ascii=False, indent=2)
    print("Saved moves.json successfully!")

if __name__ == '__main__':
    scrape()
